### DISCUSSION TITLE
Replace clsx with Tailwind-Merge

### DISCUSSION THREAD

**Discussion by onemen** [epicweb-dev/epic-stack#117] [Ideas]
@kentcdodds

I've watched this video, and this package is looking good.

what do you think ?

here a Youtube video from @simonswiss about [Tailwind-Merge](https://www.youtube.com/watch?v=tfgLd5ZSNPc)
https://github.com/dcastil/tailwind-merge

**Comment by kentcdodds** [comment]
I've been thinking about that. I'm a little concerned that it's 21kb of JS, but I do think that it's worth its weight.

**Comment by onemen** [comment]
At the moment I don't think we that `epic-stack` have any class name duplication, so we don't need the benefit of merge

**Reply by kentcdodds** [reply]
True, but we could put it in place for our form components so people could start passing class names to those without having to worry about this problem.

**Comment by onemen** [comment]
maybe we can use both `tailwind-merge` and `clsx`

`clsx` in getButtonClassName and `tailwind-merge` in the component:
```tsx
			className={twMerge(
				getButtonClassName({ size, variant }),
				'flex justify-center gap-4',
				props.className,
			)}
```

**Reply by kentcdodds** [reply]
I think tailwind-merge has a join utility that functions just like clsx so I think we'd drop clsx if we did this.

**Reply by rajeshg** [reply]
twJoin - https://github.com/dcastil/tailwind-merge/blob/v1.13.0/docs/when-and-how-to-use-it.md

```js
className={twMerge(twJoin(
    getButtonClassName({ size, variant }), 
    'flex justify-center gap-4', 
    props.className
))}		

``` 

May be we can add a utils function like this (reference: https://github.com/shadcn/ui/blob/main/apps/www/lib/utils.ts)
```js
export function mergeJoin(...inputs: ClassValue[]) {
  return twMerge(twJoin(inputs))
}

```

**Reply by kentcdodds** [reply]
100%

Definitely willing to accept a PR for this.

**Reply by kentcdodds** [reply]
Though, honestly I'm pretty sure we could just use `twMerge` and not bother with `twJoin` here. I may be missing something though.

**Reply by courdek** [reply]
The key difference between `twMerge/twJoin` and `clsx` is that the former [does not support the `object` syntax](https://github.com/dcastil/tailwind-merge/discussions/137#discussioncomment-3481605). Otherwise `twMerge` could be used by itself, although it is slower than both `twJoin` and `clsx` because of the conflict resolution.
